generator/feature_maker_base.cpp

```diff
@@ -0,0 +1,75 @@
+#include "generator/feature_maker_base.hpp"
+
+#include "generator/osm_element.hpp"
+
+namespace generator
+{
+FeatureMakerBase::FeatureMakerBase(cache::IntermediateDataReader & holder) : m_holder(holder) { }
+
+bool FeatureMakerBase::Add(OsmElement & p)
+{
+  FeatureParams params;
+  ParseParams(params, p);
+  switch (p.type)
+  {
+  case OsmElement::EntityType::Node:
+    return BuildFromNode(p, params);
+  case OsmElement::EntityType::Way:
+    return BuildFromWay(p, params);
+  case OsmElement::EntityType::Relation:
+    return BuildFromRealtion(p, params);
+  default:
+    return false;
+  }
+}
+
+size_t FeatureMakerBase::Size() const
+{
+  return m_queue.size();
+}
+
+bool FeatureMakerBase::Empty() const
+{
+  return m_queue.empty();
+}
+
+bool FeatureMakerBase::GetFeature(FeatureBuilder1 & feature)
+{
+  if (m_queue.empty())
+    return false;
+
+  feature = std::move(m_queue.front());
+  m_queue.pop();
+  return true;
+}
+
+
+void TransformAreaToPoint(FeatureBuilder1 & feature)
+{
+  auto const center = feature.GetGeometryCenter();
+  auto const id = feature.GetLastOsmId();
```

Comment:
а почему Last, а не MostGeneric?

Reply:
я так понимаю, что если у нас Area была relation-ом, то id relation-а это MostGenericOsmId

Reply:
так было в оригинальном коде, иногда прошу поглядывать в удаленный файл translator_planet.cpp

Reply:
код переехавший из translator_planet это не единственное место где используется этот метод

Reply:
up

Reply:
если конкретно в карте мира теперь другие ID у popular attractions

Reply:
это нужно повнимательнее глянуть, до этого всегда с этим вызовом работала сборка карт

Reply:
и вообще, если у нас мультиполигональный релейшен, то что будет в GetLastOsmId? и почему мы хотим именно его брать? я не понимаю пока что там и почему мы именно Last берём

Reply:
не знаю, почему авторы испольовали Last, наверное потому что это реализация проще и быстрее. я тоже за MostGenericOsmId, исследую этот вопрос потом, сейчас уже  сделал сборку карт без него

Reply:
хорошо, спасибо!

Reply:
как я понял GetLastOsmId это id самой фичи, которая может содержать в себе другие node/way/relation
предлагаю переименовать на, например, GetRootOsmId() и удалить GetFirstOsmId() и GetMostGenericOsmId()

Reply:
вообще OsmId самой фичи это MostGenericOsmId

возможно (надо это проверить) процесс создания FeatureBuilder устроен так, что он ещё заодно оказывается и Last, предлагаю с этим в отдельном реквесте разбираться
```
/// @returns an id of the most general element: node's one if there is no area or relation,
/// area's one if there is no relation, and relation id otherwise.
base::GeoObjectId GetMostGenericOsmId() const;
```